celery/canvas.py

```diff
@@ -395,23 +395,37 @@ def __or__(self, other):
             other = maybe_unroll_group(other)
             if isinstance(self, _chain):
                 # chain | group() -> chain
-                sig = self.clone()
-                sig.tasks.append(other)
-                return sig
+                tasks = [t.clone() for t in self.tasks]
+                link_error = self.options.get('link_error', [])
+                for sig in link_error:
+                    for task in tasks:
+                        task.on_error(sig)
+                return _chain(seq_concat_item(tasks, other), app=self._app)
             # task | group() -> chain
             return _chain(self, other, app=self.app)
 
         if not isinstance(self, _chain) and isinstance(other, _chain):
             # task | chain -> chain
-            return _chain(
-                seq_concat_seq((self,), other.tasks), app=self._app)
+            tasks = [t.clone() for t in other.tasks]
+            link_error = other.options.get('link_error', [])
+            for sig in link_error:
+                for task in tasks:
+                    task.on_error(sig)
+            return _chain(seq_concat_seq((self,), tasks), app=self._app)
         elif isinstance(other, _chain):
             # chain | chain -> chain
-            sig = self.clone()
-            if isinstance(sig.tasks, tuple):
-                sig.tasks = list(sig.tasks)
-            sig.tasks.extend(other.tasks)
-            return sig
+            # assign chain's link_error sugnatures to each chain's task
+            tasks = [t.clone() for t in self.tasks]
+            link_error = self.options.get('link_error', [])
+            for sig in link_error:
+                for task in tasks:
+                    task.on_error(sig)
+            other_tasks = [t.clone() for t in other.tasks]
```

Comment:
Seems this is a repeated pattern, the more DRY approach would be to use a function for these 5 occurrences.

Reply:
I've though about it when applying your suggestion :)
Wait for one more commit.